generate/types.go

```diff
@@ -1,412 +1,188 @@
-	if err != nil {
-		return "", err
-	}
-
-	baseType, err := g.baseTypeForOperation(operation.Operation)
-	if err != nil {
-		return "", errorf(operation.Position, "%v", err)
-	}
-
-	description := fmt.Sprintf("%v is returned by %v on success.", name, operation.Name)
-
-	builder := &typeBuilder{generator: g}
-	err = builder.writeTypedef(
-		name, operation.Name, baseType, operation.Position, fields, queryOptions, description)
-	return name, err
-}
-
-var builtinTypes = map[string]string{
-	// GraphQL guarantees int32 is enough, but using int seems more idiomatic
-	"Int":     "int",
-	"Float":   "float64",
-	"String":  "string",
-	"Boolean": "bool",
-	"ID":      "string",
-}
-
-func (g *generator) typeName(prefix string, typ *ast.Definition) (name, nextPrefix string) {
-	typeGoName := upperFirst(typ.Name)
-	if typ.Kind == ast.Enum || typ.Kind == ast.InputObject {
-		// If we're an enum or an input-object, there is only one type we
-		// will ever possibly generate for this type, so we don't need any
-		// of the qualifiers.  This is especially helpful because the
-		// caller is very likely to need to reference these types in their
-		// code.
-		return typeGoName, typeGoName
-	}
-
-	name = prefix
-	if !strings.HasSuffix(prefix, typeGoName) {
-		// If the field and type names are the same, we can avoid the
-		// duplication.  (We include the field name in case there are
-		// multiple fields with the same type, and the type name because
-		// that's the actual name (the rest are really qualifiers); but if
-		// they are the same then including it once suffices for both
-		// purposes.)
-		name += typeGoName
-	}
-
-	if typ.Kind == ast.Interface || typ.Kind == ast.Union {
-		// for interface/union types, we do not add the type name to the
-		// name prefix; we want to have QueryFieldType rather than
-		// QueryFieldInterfaceType.  So we just use the input prefix.
-		return name, prefix
-	}
-
-	// Otherwise, the name will also be the prefix for the next type.
-	return name, name
-}
-
-func (g *generator) getTypeForInputType(opName string, typ *ast.Type, options, queryOptions *GenqlientDirective) (string, error) {
-	// Sort of a hack: case the input type name to match the op-name.
-	name := matchFirst(typ.Name(), opName)
-	builder := &typeBuilder{generator: g}
-	// note prefix is ignored here (see generator.typeName)
-	err := builder.writeType(name, "", typ, selectionsForInputType(g, typ, queryOptions), options)
-	return builder.String(), err
-}
+// goType represents a type for which we'll generate code.
+type goType interface {
+	// WriteDefinition writes the code for this type into the given io.Writer.
+	//
+	// TODO(benkraft): Some of the implementations might now benefit from being
+	// converted to templates.
+	WriteDefinition(io.Writer, *generator) error
+
+	// Reference returns the Go name of this type, e.g. []*MyStruct, and may be
+	// used to refer to it in Go code.
+	Reference() string
```

Comment:
What do you think about naming this "TypeName" instead of "Reference". "Reference" sounds like this should always be a pointer.

Reply:
The problem is it's not always a type-name either since it might be `[]*MyStruct`.  But maybe that's at least closer.